mycroft/tts/cache.py

```diff
@@ -0,0 +1,292 @@
+)
+from mycroft.util.log import LOG
+
+
+def _get_mimic2_audio(sentence: str, url: str) -> Tuple[bytes, str]:
+    """Use the Mimic2 API to retrieve the audio for a sentence.
+
+    Arguments:
+        sentence: The sentence to be cached
+    """
+    LOG.debug("Retrieving Mimic2 audio for sentence \"{}\'".format(sentence))
+    mimic2_url = url + parse.quote(sentence) + '&visimes=True'
+    response = requests.get(mimic2_url)
+    response_data = response.json()
+    audio = base64.b64decode(response_data["audio_base64"])
+    phonemes = response_data["visimes"]
+
+    return audio, phonemes
+
+
+def hash_sentence(sentence: str):
+    """Convert the sentence into a hash value used for the file name
+
+    Arguments:
+        sentence: The sentence to be cached
+    """
+    encoded_sentence = sentence.encode("utf-8", "ignore")
+    sentence_hash = hashlib.md5(encoded_sentence).hexdigest()
+
+    return sentence_hash
+
+
+class AudioFile:
+    def __init__(self, cache_dir: Path, sentence_hash: str, file_type: str):
+        self.name = f"{sentence_hash}.{file_type}"
+        self.path = cache_dir.joinpath(self.name)
+
+    def save(self, audio: bytes):
+        """Write a TTS cache file containing the audio to be spoken.
+
+        Arguments:
+            audio: TTS inference of a sentence
+        """
+        try:
+            with open(self.path, "wb") as audio_file:
+                audio_file.write(audio)
+        except Exception:
+            LOG.exception("Failed to write {} to cache".format(self.name))
+
+
+class PhonemeFile:
+    def __init__(self, cache_dir: Path, sentence_hash: str):
+        self.name = f"{sentence_hash}.pho"
+        self.path = cache_dir.joinpath(self.name)
+
+    def load(self) -> str:
+        """Load phonemes from cache file."""
+        phonemes = None
+        if self.path.exists():
+            try:
+                with open(self.path) as phoneme_file:
+                    phonemes = phoneme_file.read().strip()
+            except Exception:
+                LOG.exception("Failed to read phoneme from cache")
+
+        return phonemes
+
+    def save(self, phonemes):
+        """Write a TTS cache file containing the phoneme to be displayed.
+
+        Arguments:
+            phonemes: instructions for how to make the mouth on a device move
+        """
+        if type(phonemes) == str:
+            rec = phonemes
+        else:
+            rec = json.dumps(phonemes)
+        try:
+            with open(self.path, "w") as phoneme_file:
+                phoneme_file.write(rec)
+        except Exception:
+            LOG.exception("Failed to write {} to cache".format(self.name))
+
+
+class TextToSpeechCache:
+    """Class for all persistent and temporary caching operations."""
+    def __init__(self, tts_config, tts_name, audio_file_type):
+        self.config = tts_config
+        self.tts_name = tts_name
+        self.persistent_cache_dir = Path(tts_config.get("preloaded_cache"))
```

Comment:
This config doesn't exist for all tts's and throws an error when missing.

Reply:
Good catch.  Will fix.